### ISSUE TITLE
Possibility to listen to selection change events

### ISSUE THREAD

**Issue by bdlukaa** [flutter/flutter#101780]
## Use case

I want to create a mention system (instagram-like), but I am currently not able to listen to the selection events and check if the cursor position has updated, and update the app state accordingly.

## Proposal

Add a listener that is able to listen to selection changes, or even propagate those selection changes on the already existent `controller.addListener` callback itself.

It'd be something like:
```dart
final controller = TextEditingController();

controller.selection.addListener(() {
  debugPrint('selection changed');
});
```

or even better:

```dart
final controller = TextEditingController();

controller.selection.onChanged((old) {
  print('old: $old , new ${controller.selection}');
});
```

**Comment by bdlukaa** [comment]
@chunhtai I tried on the latest stable and it does not notify on controller.addListener

**Comment by maheshmnj** [comment]
Hi @bdlukaa, Thanks for filing the issue. I see that `controller.addListener` does notify on cursor position change, text change, or text selection change. Please take a look at the below code sample 

<details>
<summary>flutter doctor -v</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late TextEditingController textEditingController;

  @override
  void initState() {
    super.initState();
    textEditingController = TextEditingController();
    textEditingController.addListener(() {
      print('change in controller text ${textEditingController.selection}');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: TextField(
        controller: textEditingController,
      )),
    );
  }
}

```
</details> 

<details>
<summary>logs</summary>

```
mahesh@Maheshs-MacBook-Air-M1 new_app % flutter run -d macos
Launching lib/main.dart on macOS in debug mode...
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:macOS, arch:x86_64, id:6A745C96-D83E-5C8D-8EE9-5958E598001B }
{ platform:macOS, name:Any Mac }
Building macOS application...                                          ⣷        
Syncing files to device macOS...                                   174ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on macOS is available at: http://127.0.0.1:62373/xvpA4cfNG-Q=/

The Flutter DevTools debugger and profiler on macOS is available at: http://127.0.0.1:9101?uri=http://127.0.0.1:62373/xvpA4cfNG-Q=/
flutter: change in controller text TextSelection.collapsed(offset: 0, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 1, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 2, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 3, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 4, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 5, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 6, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 7, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 8, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 9, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 10, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 11, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 12, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 13, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 14, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 17, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 1, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 0, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 1, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 2, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 3, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 4, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 5, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 6, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 7, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 8, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 9, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 10, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 11, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 12, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 13, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 14, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 17, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 20, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 19, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 18, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 17, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 16, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 15, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 14, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 13, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 12, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 11, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 10, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 9, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 8, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 7, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 6, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 5, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 4, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 3, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 2, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 1, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 0, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 1, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 2, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 3, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 4, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 5, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 9, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 1, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 2, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 3, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 4, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 5, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 6, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 7, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 8, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 9, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 10, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 11, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 12, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 13, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 14, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 1, extentOffset: 15, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 19, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 20, affinity: TextAffinity.upstream, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 19, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 17, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 16, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 15, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 14, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 13, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 12, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 11, isDirectional: false)
flutter: change in controller text TextSelection(baseOffset: 20, extentOffset: 10, isDirectional: false)
flutter: change in controller text TextSelection.collapsed(offset: 10, affinity: TextAffinity.downstream, isDirectional: false)
```  
</details>

**Comment by bdlukaa** [comment]
True! Something was wrong on my end. Thanks <3
